continuous-integration/readme.md

```diff
@@ -1,3 +1,227 @@
+- [ ] **No IDE Dependencies**
+  - It's essential to have a master build that's runnable through standalone scripts and not dependent on a particular IDE. Build pipeline targets can be triggered locally on their desktops through their IDE of choice. The build process should maintain enough flexibility to run within a CI server as well. As an example, dockerizing your build process offers this level of flexibility as VSCode and IntelliJ supports [docker plugin](https://code.visualstudio.com/docs/containers/overview) extensions.
+
+## Build Environment Dependencies
+
+- [ ] **Automated local environment setup**
+  - We encourage maintaining a consistent developer experience for all team members. There should be a central automated manifest / process that streamlines the installation and setup of any software dependencies. This way developers can replicate the same build environment locally as the one running on a CI server.
+  - Build automation scripts often require specific software packages and version pre-installed within the runtime environment of the OS. This presents some challenges as build processes typically version lock these dependencies.
+  - All developers on the team should be able to emulate the build environment from their local desktop regardless of their OS.
+  - Developers should trigger the automated environment setup task(s) through a single command.
+  - Well established software packaging tools like Docker, Maven, npm, etc should be considered when designing your build automation tool chain.
+- [ ] **Document local setup**
+  - The setup process for setting up a local build environment should be well documented and easy for developers to follow.
+  
+## Infrastructure as Code
+
+Manage as much of the following as possible, as code:
+
+- Configuration Files
+- Configuration Management
+- Secret Management
+- Cloud Resource Provisioning
+- Role Assignments
+- Load Test Scenarios
+- Availability Alerting / Monitoring Rules and Conditions
+
+Decoupling infrastructure from the application codebase simplifies engineering teams move to cloud native applications.
+
+### Sample DevOPS Workflow using Terraform and Cobalt
+
+![image](https://user-images.githubusercontent.com/7635865/76626035-652eda80-6506-11ea-8870-6070365f10d6.png)
+
+### Why
+
+- Repeatable and auditable changes to infrastructure make it easier to roll back to known good configurations and to rapidly expand to new stages and regions without having to hand-wire cloud resources
+- Battle tested and templatized IAC reference projects like [Cobalt](https://github.com/microsoft/cobalt) and [Bedrock](https://github.com/microsoft/bedrock) enable more business teams deploy secure and scalable solutions at a much more rapid pace
+- Simplify “lift and shift” scenarios by abstracting the complexities of cloud-native computing away from application developer teams.
+
+### IAC DevOPS: Operations by Pull Request
```

Comment:
Is the intent to recommend GitOps for everything? What about resource types that are not able to pull?

Reply:
Somewhat, yes. My original intent is that both system and infrastructure state are reproducible through code and a devops workflow. Things like ssl certificates and ssh key pairs would be managed outside of this process typically by security / network admins. Besides certs and keys, are there other resource types that you can foresee which couldn't be pulled from a repo?

Reply:
Unclear if are discussing [GitOps](https://medium.com/@timfpark/highly-effective-kubernetes-deployments-with-gitops-c7a0354f1446) or simply changes that are promoted (via push or pull) to one or many target environments in response to a git commit.

The former is too opinionated, in my opinion, to recommend as a general practice.

Reply:
I believe @xtophs was referring to b)?